lib/zk_recipes/cache.rb

```diff
@@ -4,20 +4,35 @@ module ZkRecipes
   class Cache
     class Error < StandardError; end
     class PathError < Error; end
+    class StateError < Error; end
 
     AS_NOTIFICATION = "cache.zk_recipes"
+    AS_NOTIFICATION_DIRECTORY = "directory_cache.zk_recipes"
     USE_DEFAULT = Object.new
 
+    STATIC = Object.new
+    private_constant :STATIC
+
     def initialize(logger: nil, host: nil, timeout: nil, zk_opts: {})
-      @cache = Concurrent::Map.new
-      @latch = Concurrent::CountDownLatch.new
       @logger = logger
-      @pending_updates = Concurrent::Hash.new # Concurrent::Map does not implement #reject!
-      @registerable = true
-      @registered_values = Concurrent::Map.new
       @session_id = nil
-      @watches = Concurrent::Map.new
       @zk = nil
+      @latch = Concurrent::CountDownLatch.new
+
+      # Concurrent::Hash preserves insertion order, but you can't mutate the hash while iterating
+      @pending_updates = Concurrent::Map.new
+
+      # These paths need to be registered *before* setup_callbacks
+      @static_paths = {}
+      @static_watches = Concurrent::Map.new
+
+      # These paths need to be registered *before* setup_callbacks
+      @directory_paths = {}
```

Comment:
Guessing that both of these are hashes because of the "before" part?

Reply:
Ya, I _think_ that since they only get written to before the cache client is setup, it should be threadsafe to read from them. I could `freeze` them to make sure that's the case I suppose.